overlord/snapstate/component.go

```diff
@@ -209,3 +211,130 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
 
 	return installSet, nil
 }
+
+type RemoveComponentsOpts struct {
+	RefreshProfile bool
+}
+
+// RemoveComponents returns a taskset that removes the components in compName
+// that belog to snapName.
+func RemoveComponents(st *state.State, snapName string, compName []string, opts RemoveComponentsOpts) ([]*state.TaskSet, error) {
+	var snapst SnapState
+	err := Get(st, snapName, &snapst)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+	if !snapst.IsInstalled() {
+		return nil, &snap.NotInstalledError{Snap: snapName, Rev: snap.R(0)}
+	}
+
+	info, err := snapst.CurrentInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	var setupSecurity *state.Task
+	if opts.RefreshProfile {
+		revisionStr := fmt.Sprintf(" (%s)", info.Revision)
+		setupSecurity = st.NewTask("setup-profiles",
+			fmt.Sprintf(i18n.G("Setup snap %q%s security profiles"), snapName, revisionStr))
+	}
+
+	var tss []*state.TaskSet
+	for _, comp := range compName {
+		cref := naming.NewComponentRef(snapName, comp)
+		compst := snapst.CurrentComponentState(cref)
+		if compst == nil {
+			return nil, &snap.ComponentNotInstalledError{
+				NotInstalledError: snap.NotInstalledError{
+					Snap: info.InstanceName(),
+					Rev:  info.Revision,
+				},
+				Component: comp,
+				CompRev:   snap.R(0),
+			}
+		}
+		ts, err := removeComponentTasks(st, compst, info, setupSecurity)
+		if err != nil {
+			return nil, err
+		}
+		ts.JoinLane(st.NewLane())
```

Comment:
do we have a test that checks what happens if only one of the removal fails, as they are different lanes the other will not be undone but then I'm not sure we will run the setup-profiles as it is waiting on everything?

Reply:
I started to write a test and that helped me see that there was a bug, as setup-profiles needs to happen before discard-component as the latter cannot be undone (it removes the component file).

Also, after thinking about this, maybe we should make removal of multiple components for a snap transactional anyway, having a situation where you are half the way is not really desirable as we might have dependencies betwee components that are not expected for snaps (we are already seing that for some use cases). So for the moment I've removed the call to JoinLanes, but let me know what you think.

Reply:
yes, that's fine